# ការរៀបចំទិន្នន័យ

[មូលដ្ឋានកំណត់ត្រាដើមពី *វិទ្យាសាស្ត្រទិន្នន័យ៖ ការណែនាំអំពីការរៀនម៉ាស៊ីនសម្រាប់ Python និង Machine Learning Studio ដោយ Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## ការស្វែងយល់ព័ត៌មាន `DataFrame`

> **គោលបំណងរៀន៖** នៅចុងផ្នែកនេះ អ្នកគួរតែមានភាពងាយស្រួលក្នុងការស្វែងរកព័ត៌មានទូទៅអំពីទិន្នន័យដែលបានរក្សាទុកក្នុង pandas DataFrames។

ពេលដែលអ្នកបានផ្ទុកទិន្នន័យរបស់អ្នកចូលក្នុង pandas វានឹងមានទំរង់ជា `DataFrame` ជាច្រើនច្រើនជាងមួយ។ ទោះយ៉ាងណា ប្រសិនបើទិន្នន័យក្នុង `DataFrame` របស់អ្នកមានជួរដេក 60,000 និងជួរឈរ 400 តើអ្នកដឹងដូចម្តេចក្នុងការចាប់ផ្តើមយល់អំពីអ្វីដែលអ្នកកំពុងធ្វើ? សំណាងល្អ pandas ផ្តល់ឧបករណ៍ងាយស្រួលមួយចំនួនសម្រាប់មើលព័ត៌មានទូទៅអំពី `DataFrame` ដោយរហ័ស លើសពីស្រទាប់ជួរដំបូង និងចុងក្រោយ។

ដើម្បីស្វែងយល់ពីមុខងារនេះ អ្នកនឹងនាំចូលបណ្ណាល័យ Python scikit-learn ហើយប្រើសំណុំទិន្នន័យដ៏ល្បីមួយ ដែលអ្នកវិទ្យាសាស្ត្រទិន្នន័យគ្រប់រូបបានឃើញរាប់រ៉ុកដង៖ សំណុំទិន្នន័យ *Iris* របស់ជីវសាស្ត្រ ប៊ូលរីស្ត៍ Ronald Fisher ដែលបានប្រើក្នុងអត្ថបទរបស់គាត់ឆ្នាំ 1936 "ការប្រើវាស់វែងច្រើនក្នុងបញ្ហាតាក់ស៊ូម៉ិក":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
យើងបានផ្ទុក Dataset Iris ទៅក្នុងអថេរ `iris_df`។ មុននឹងចូលទៅក្នុងទិន្នន័យ វានឹងមានប្រយោជន៍ដើម្បីដឹងចំនួនចំណុចទិន្នន័យដែលយើងមាន និងទំហំទាំងមូលរបស់ dataset។ វាមានប្រយោជន៍ក្នុងការមើលបរិមាណទិន្នន័យដែលយើងកំពុងដោះស្រាយ។


In [2]:
iris_df.shape

(150, 4)

ដូច្នេះ យើងកំពុងដំណើរការជាតួអក្សរ 150 និងជួរឈរចំនួន 4 នៃទិន្នន័យ។ តួអក្សរនីមួយៗតំណាងឱ្យចំណុចទិន្នន័យមួយ និងជួរឈរនីមួយៗតំណាងឱ្យលក្ខណៈពិសេសមួយដែលពាក់ព័ន្ធនឹងស៊ុមទិន្នន័យ។ ដូច្នេះមូលដ្ឋានទាំងអស់មានចំណុចទិន្នន័យ 150 ដែលមានលក្ខណៈពិសេស 4 គ្រប់គ្រាន់។

`shape` នៅទីនេះគឺជាគុណលក្ខណៈនៃស៊ុមទិន្នន័យ មិនមែនជាអនុគមន៍ទេ ដូច្នេះវាមិនបញ្ចប់ជាមួយគូវគ្គនៃសញ្ញាវិថីទេ។


### `DataFrame.columns`
ឥឡូវនេះយើងមកពិនិត្យកាន់តែជ្រៅទៅលើជួរឈរទិន្នន័យចំនួន ៤។ តើមួយៗក្នុងនោះតំណាងអ្វីខ្លះ? ទ្រព្យសម្បត្តិ `columns` នឹងផ្តល់ឲ្យ​យើងឈ្មោះនៃជួរឈរនៅក្នុង dataframe។


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

ដូចដែលយើងអាចមើលឃើញ មានជួរឈរ​បួន(4)។ លក្ខណៈ `columns` បង្ហាញឲ្យយើងដឹងពីឈ្មោះនៃជួរឈរនិងគ្មានអ្វីផ្សេងទៀត។ លក្ខណៈនេះមានសារៈសំខាន់នៅពេលដែលយើងចង់កំណត់លក្ខណៈពិសេសដែលឯកសារទិន្នន័យមាន។


### `DataFrame.info`
ចំនួនទិន្នន័យ(ដែលផ្តល់ដោយគុណលក្ខណៈ `shape`) និងឈ្មោះនៃលក្ខណៈឬជួរឈរដែលផ្តល់ដោយគុណលក្ខណៈ `columns` បានប្រាប់យើងអំពីសំណុំទិន្នន័យមួយ។ ឥលូវនេះ យើងចង់ធ្វើយកចូលទៅកាន់សំណុំទិន្នន័យខាងក្នុងជាងមុន។ មុខងារ `DataFrame.info()` គឺមានប្រយោជន៍យ៉ាងខ្លាំងសម្រាប់ការនេះ។


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


នៅទីនេះ យើងអាចធ្វើការប៉ាន់ស្មានខ្លះៗបាន៖
1. ប្រភេទទិន្នន័យនៃជួរដេកនីមួយៗ៖ ក្នុងឯកសារទិន្នន័យនេះ ទិន្នន័យទាំងអស់ត្រូវបានផ្ទុកជាចំនួនទសភាគចុចហ្រ្វ្លូត 64 ប៊ីត។
2. លេខនៃតម្លៃមិនមែនជាសូន្យ៖ ការដោះស្រាយតម្លៃសូន្យគឺជាជំហានសំខាន់មួយក្នុងការរៀបចំទិន្នន័យ។ វានឹងត្រូវបានដោះស្រាយនៅពេលក្រោយក្នុងសៀវភៅកំណត់សម្គាល់។


### DataFrame.describe()
សន្មត់ថាយើងមានទិន្នន័យច្រើនជាលេខនៅក្នុង dataset របស់យើង។ ការគណនាស្ថិតិមួយប៉ុណ្ណោះ ដូចជាមធ្យមភាព មេឌៀន ចំណុចបួនជាប់លំដាប់ ល។ អាចធ្វើបានលើ១១ាយជួរឈរ ដោយឡែក។ មុខងារ `DataFrame.describe()` នាំឱ្យយើងទទួលបានសង្ខេបស្ថិតិ នៃជួរឈរលេខនានានៅក្នុង dataset ។


In [5]:
iris_df.describe()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
count,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333
std,0.828066,0.435866,1.765298,0.762238
min,4.300000,2.000000,1.000000,0.100000
25%,5.100000,2.800000,1.600000,0.300000
50%,5.800000,3.000000,4.350000,1.300000
75%,6.400000,3.300000,5.100000,1.800000
max,7.900000,4.400000,6.900000,2.500000


លទ្ធផលខាងលើបង្ហាញពីចំនួនតុល្យឯកតាទិន្នន័យសរុប មធ្យម គោលឆ្នាំត្រង់ កម្រិតអប្បបរមា ជួរចុះក្រោម(២៥%), មេឌាន(៥០%), ជួរឧទ្ធម្ភ(៧៥%) និងតម្លៃអតិបរមារបស់ក្រុមជួរឈរ។


### `DataFrame.head`
ជាមួយនឹងមុខងារ និងគុណលក្ខណៈទាំងអស់ខាងលើ យើងបានទទួលបានទិដ្ឋភាពកំណត់កម្រិតកំពូលនៃសំណុំទិន្នន័យ។ យើងដឹងថាមានចំនួនចំណុចទិន្នន័យប៉ុន្មាន មានលក្ខណៈប៉ុន្មាន ប្រភេទទិន្នន័យរបស់លក្ខណៈនីមួយៗ និងចំនួនតម្លៃមិនមែន null សម្រាប់លក្ខណៈនីមួយៗ។

ឥឡូវនេះពេលវេលាដើម្បីមើលទិន្នន័យផ្ទាល់។ យើងមកមើលថាតួរដេកមួយចំនួនដំបូង(ចំណុចទិន្នន័យមួយចំនួនដំបូង) នៃ `DataFrame` របស់យើងមើលទៅដូចម្តេច:


In [6]:
iris_df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2


ជា​​​ចេញ​ផល​នៅ​ទីនេះ​យើង​អាច​ឃើញ​ទន្សាយ(៥)​​ចំនួន​​នៃ​ទិន្នន័យ។ ប្រសិន​បើយើង​មើល​នៅ​លេខសន្ទស្សន៍​ខាងឆ្វេង យើង​រក​ឃើញ​ថា​ទាំងនេះ​ជា​បន្ទាត់​ប្រាំ​ដំបូង។


### លំហាត់៖

ពីឧទាហរណ៍ដែលបានផ្តល់ខាងលើ វាច្បាស់ហើយថា ជាដំណើរការស្វ័យប្រវត្តិ `DataFrame.head` បញ្ចូនជួរដេកដំបូងប្រាំនៃ `DataFrame`។ នៅក្នុងក្រឡាចត្រង់កូដខាងក្រោម អ្នកអាចស្វែងរកវិធីបង្ហាញជួរដេកច្រើនជាងប្រាំបានទេ?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
វិធីមួយផ្សេងទៀតនៃការមើលទិន្ន័យអាចមកពីចុង (ជំនួសពីការចាប់ផ្តើម)។ ផ្នែកបំផុតនៃ `DataFrame.head` គឺ `DataFrame.tail` ដែលត្រឡប់មកវិញជួរដកចុងបញ្ចប់បួនចុងក្រោយនៃ `DataFrame` ៖


In [8]:
iris_df.tail()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3
149,5.9,3.0,5.1,1.8


នៅក្នុង​អនុវត្តន៍​វិជ្ជាមាន​ប្រយោជន៍​នៅ​ក្នុង​ការ​អាច​ពិនិត្យមើល​ជួរដេក​ដំបូង​ជាច្រើន ឬ​ជួរដេកចុងក្រោយ​ជាច្រើន​របស់ `DataFrame` បានយ៉ាងងាយស្រួល ពិសេសនៅពេលដែលអ្នកកំពុងស្វែងរក​តម្លៃឆ្លងច្រាំង​នៅក្នុងឈុតទិន្នន័យដែលរៀបចំតាមលំដាប់។

មុខងារ និងគុណលក្ខណៈទាំងអស់ដែលបានបង្ហាញខាងលើជាមួយនឹងឧទាហរណ៍កូដបានជួយឲ្យយើងទទួលបានទិដ្ឋភាពនិងអារម្មណ៍លើទិន្នន័យ។

> **ចំណាំ៖** ទោះបីជាត្រឹមតែ​មើលព័ត៌មានអំពី Metadata អំពីព័ត៌មាន​នៅក្នុង DataFrame ឬតម្លៃដំបូងនិងចុងក្រោយប៉ុណ្ណោះ អ្នកក៏អាចទទួលបានយោបល់ភ្លាមៗអំពីទំហំ រាងកាយ និងមាតិកា​នៃទិន្នន័យដែលអ្នកកំពុងដោះស្រាយ។


### ទិន្នន័យដែលខ្វះ
ចូលចិត្តយើងចូលទៅក្នុងទិន្នន័យដែលខ្វះ។ ទិន្នន័យដែលខ្វះមានកើតឡើងពេលដែលគ្មានតម្លៃបានរក្សាទុកក្នុងខ្នាតខ្លះៗ។ 

យើងឧទាហរណ៍៖ គេនិយាយថាមនុស្សម្នាក់ព្រមព្រៀងពីទម្ងន់របស់គាត់ ហើយមិនបំពេញវាលទម្ងន់នៅក្នុងការស្ទង់មតិ។ បន្ទាប់មក តម្លៃទម្ងន់សម្រាប់មនុស្សនោះនឹងខ្វះ។ 

ភាគច្រើននៅក្នុងឯកសារទិន្នន័យពិភពលោកពិត តម្លៃដែលខ្វះនឹងកើតមាន។

**របៀបដែល Pandas ដោះស្រាយទិន្នន័យដែលខ្វះ**


Pandas ដោះស្រាយតម្លៃដែលខ្វះពីរផ្លូវ។ ជាលេខមួយដែលអ្នកបានឃើញមុននេះនៅក្នុងផ្នែកមុនៈ `NaN` ឬ Not a Number។ វាជាតម្លៃពិសេសមួយដែលជាផ្នែកនៃការបញ្ជាក់ IEEE floating-point ហើយវាត្រូវបានប្រើសំរាប់បង្ហាញតែតម្លៃ floating-point ដែលខ្វះប៉ុណ្ណោះ។

សំរាប់តម្លៃដែលខ្វះក្រៅពី float, pandas ប្រើវត្ថុ Python `None`។ បើទោះបីវាហាក់ដូចជារអិលថ្លៃថាអ្នកនឹងជួបប្រទៈតម្លៃពីរប្រភេទខុសគ្នាដែលនិយាយអំពីអ្វីដែលដូចគ្នាពិតប្រាកដ ក៏មានហេតុខាងក្រោយដែលមានមូលដ្ឋានល្អសំរាប់ការប្រើប្រាស់នេះ ហើយពិតជាក្នុងការអនុវត្ត វagde pandas អាចផ្ដល់ការប្រព្រឹត្តល្អសម្រាប់ករណីភាគច្រើន។ ទោះជាយ៉ាងណា ទាំង `None` និង `NaN` មានការកំណត់ដែលអ្នកត្រូវចងចាំពាក់ព័ន្ធនឹងរបៀបដែលពួកវាអាចត្រូវបានប្រើ។


### `None`: ទិន្នន័យដែលខ្វះមិនមែនជាប្រភេទ float
ដោយសារតែ `None` មកពី Python វាមិនអាចប្រើបានក្នុង NumPy និង pandas arrays ដែលមិនមែនជាប្រភេទទិន្នន័យ `'object'` ទេ។ ចងចាំថា NumPy arrays (និងរចនាសម្ព័ន្ធទិន្នន័យនៅក្នុង pandas) អាចមានបានតែក្នុងលក្ខណៈប្រភេទទិន្នន័យតែមួយប៉ុណ្ណោះ។ នេះគឺជាវត្ថុដែលផ្តល់ឱ្យពួកវាមានអំណាចយ៉ាងសំខាន់សម្រាប់ការដំណើរការទិន្នន័យក្នុងមាឌធំ និងការគណនា ប៉ុន្តែមិនអាចបត់បែនបានយូរទេ។ អារេបែបនេះត្រូវប្រើការកំណត់ប្រភេទទិន្នន័យទៅ “lowest common denominator” វា​គឺជាប្រភេទទិន្នន័យដែលនឹងគ្របដណ្តប់លើអ្វីគ្រប់យ៉ាងនៅក្នុងអារេ។ នៅពេលមាន `None` នៅក្នុងអារេ នោះមានន័យថាអ្នកកំពុងប្រើវាជាមួយអ объекта Python។

ដើម្បីមើលឃើញអំពីការធ្វើការ នេះ សូមពិចារណាអារេគំរូដូចតទៅ (ចំណាំ `dtype` សម្រាប់វា):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

យុទ្ធសាស្រ្តពិតប្រាកដនៃប្រភេទទិន្នន័យ upcast មានផលប៉ះពាល់ពីរជាមួយវា។ ជាលើកដំបូង ប្រតិបត្តិការនឹងត្រូវបានអនុវត្តនៅកំរិតកូដ Python ដែលបានបកប្រែ ជាជំនួសជាកូដ NumPy ដែលបានបញ្ចូល។ ពិតប្រាកដណាស់ នេះមានន័យថា ប្រតិបត្តិការណាមួយដែលពាក់ព័ន្ធជាមួយ `Series` ឬ `DataFrames` ដែលមាន `None` នៅក្នុងនឹងយឺតជាង។ គួរឱ្យមាន សំណល់សម្រាប់អ្នកប្រហែលមិនបានយល់ពីការកកើតនៃការជួបប្រទៈនេះ ទោះយ៉ាងណា សម្រាប់ទិន្នន័យធំនោះ វាអាចក្លាយទៅជាបញ្ហា។

ផលប៉ះពាល់ទីពីរមកពីលទ្ធផលទីមួយ។ ពីព្រោះ `None` មានន័យថាលេងស្ទើរ `Series` ឬ `DataFrame`s ត្រឡប់ទៅជាសកល Python ត្រង់ផ្សងព្រមវា នឹងប្រើ NumPy/pandas សរុបផ្ទៃដូចជា `sum()` ឬ `min()` លើអារេដែលមានតម្លៃ ``None`` នឹងធ្វើឲ្យមានកំហុសទូទៅ៖


In [10]:
example1.sum()

TypeError: ignored

**ចំណុចសង្ខេប៖** ការបូក (និងប្រតិបត្តិការផ្សេងៗ) រវាងលេខគ inteira និងតម្លៃ `None` គឺមិនបានកំណត់ ហើយអាចកំណត់អ្វីដែលអ្នកអាចធ្វើបានជាមួយឯកសារទិន្នន័យដែលមានវា។


### `NaN`: តម្លៃស្ពោះដែលបាត់​

ផ្ទុយទៅវិញពី `None` NumPy (ហើយក៏ pandas ផងដែរ) គាំទ្រ `NaN` សម្រាប់ប្រតិបត្តិការលឿនជាបន្ទាត់ចម្ការ និង ufuncs របស់វា។ ព័ត៌មានអាក្រក់គឺថា គណនាគណិតណាមួយដែលធ្វើលើ `NaN` ស្មើនឹងបង្កើតជា `NaN` ជានិច្ច។ ឧទាហរណ៍:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

សេចក្តីល្អ៖ ការជួសជុលតម្លៃលើអារេដែលមាន `NaN` នៅក្នុងពួកវា មិនបង្កើតកំហុសឡើយ។ សេចក្តីអាក្រក់៖ លទ្ធផលមិនមានប្រយោជន៍ស្មើគ្នាឡើយ៖


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### ការហាត់ការ:


In [11]:
# What happens if you add np.nan and None together?


សូមចងចាំ៖ `NaN` គឺសម្រាប់តម្លៃបញ្ចោទទិសចុចបន្ទាត់ហោះហើរដែលបាត់បង់តែប៉ុណ្ណោះ; មិនមាន `NaN` សម្រាប់គ្រឹះទិន្នន័យចំនួនគត់ ខ្សែអក្សរ ឬតម្លៃប៊ូលែនទេ។


### `NaN` និង `None`: តម្លៃ null ក្នុង pandas

ថ្វីត្បិតជា `NaN` និង `None` អាចមានអាកប្បកិរិយាដូចមិនដូចគ្នាក្នុងបរិមាណមួយ ក្តី pandas ត្រូវបានសិក្សាឡើងដើម្បីដោះស្រាយពួកវាដោយអាចប្ដូរប្រក់ប្រជា។ ដើម្បីមើលពីអ្វីដែលយើងមានន័យ សូមពិចារណា `Series` នៃពិន្ទុលេខអាំងតេជឺ:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### លំហាត់៖


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


ក្នុងដំណើរការសម្រាប់បម្លែងប្រភេទទិន្នន័យទៅឱ្យមានភាពស្រដៀងគ្នាក្នុង `Series` និង `DataFrame`s pandas នឹងត្រូវចូលចិត្តប្តូរតម្លៃខ្វះពី `None` ទៅ `NaN` ។ ដោយសារតែមុខងារ ဒီហើយ វាអាចជួយគិតថា `None` និង `NaN` គឺជាប្រភេទ "null" ពីរប្រភេទផ្សេងគ្នាក្នុង pandas ។ ត្រឹមត្រូវហើយ នៅមានវិធីសាស្រ្តសំខាន់មួយចំនួនដែលអ្នកនឹងប្រើដើម្បីដោះស្រាយបញ្ហាតម្លៃខ្វះក្នុង pandas ដែលបញ្ជាក់ពីគំនិតនេះនៅក្នុងឈ្មោះរបស់ពួកវា៖

- `isnull()`: បង្កើតម៉ាស Boolean បង្ហាញតម្លៃខ្វះ
- `notnull()`: ផ្ទុយពី `isnull()`
- `dropna()`: ត្រឡប់ជាកំណែត្រងនៃទិន្នន័យ
- `fillna()`: ត្រឡប់ជាកំណែចម្លងនៃទិន្នន័យដែលបំពេញឬប៉ាន់ប្រមាណតម្លៃខ្វះ

ទាំងនេះគឺជាវិធីសាស្រ្តសំខាន់ដែលត្រូវចេះល្អ និងមានភាពសឿនឆាប់រំលឹក ដូចនេះយើងចង់ពិនិត្យមើលពួកវា រៀងរាល់មួយដោយជ្រៅជាងនេះ។


### ការរកឃើញតម្លៃ null

ឥឡូវនេះ យើងបានយល់ពីសារៈសំខាន់នៃតម្លៃខ្វះខាតហើយ យើងត្រូវការរកឃើញពួកវាក្នុងទិន្នន័យរបស់យើង មុនពេលដោះស្រាយពួកវា។
ទាំង `isnull()` និង `notnull()` គឺជាវិធីសាស្ត្រចម្បងរបស់អ្នកសម្រាប់រកឃើញទិន្នន័យ null។ ទាំងពីរនេះត្រឡប់មកជា mask Boolean លើទិន្នន័យរបស់អ្នក។


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

សូមមើលលម្អិតលើលទ្ធផល។ តើមានអ្វីដែលធ្វើឲ្យអ្នកភ្ញាក់ផ្អើលទេ? ខណៈពេលដែល `0` គឺជាតម្លៃសូន្យគណិតវិទ្យា តែវាបានទទួលស្គាល់ថាជាចំនួនគត់ល្អមួយ ហើយ pandas ក៏ទទួលការពិចារណាដូច្នោះដែរ។ `''` មានកម្រិតស្មុគស្មាញបន្តិចបន្តួច។ ខណៈពេលដែលយើងប្រើវានៅផ្នែកទី ១ ដើម្បីតំណាងឲ្យតម្លៃខ្សែអក្សរចន្លោះវា តែវាបានក្លាយជាវត្ថុខ្សែអក្សរម្នាក់ និងមិនមែនជាតំណាងសម្រាប់តម្លៃសូន្យតាមដែល pandas ពិចារណាជាអនុស្សាវរីយ៍ទេ។

ឥឡូវនេះ យើងចង់បង្វិលវាពីមុខទៅក្រោយ ហើយប្រើវិធីសាស្រ្តទាំងនេះនៅក្នុងរបៀបដែលអ្នកនឹងប្រើប្រាស់ពួកវាពីរបៀបជាក់ស្តែងជាងមុន។ អ្នកអាចប្រើម៉ាស Boolean ទៅមានតួនាទីជាស៊េរី ``Series`` ឬ ``DataFrame`` indexes ដោយផ្ទាល់ ដែលអាចមានប្រយោជន៍ពេលព្យាយាមធ្វើការជាមួយតម្លៃខ្វះខាត (ឬមាន) ដែលបានបំបែកចេញ។

ប្រសិនបើយើងចង់បានចំនួនសរុបនៃតម្លៃខ្វះខាត យើងអាចធ្វើការប្រមាណបូកពីម៉ាសដែលបង្កើតដោយវិធីសាស្រ្ត `isnull()` បានណាស់។


In [19]:
example3.isnull().sum()

2

### លំហាត់៖


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**ចំណុចសំខាន់**៖ វិធីសាស្រ្ត `isnull()` និង `notnull()` ទាំងពីរផ្តល់លទ្ធផលដូចគ្នា​នៅពេលអ្នកប្រើប្រាស់ពួកវាក្នុង DataFrames៖ ពួកវាបង្ហាញលទ្ធផល និងសន្ទស្សន៍នៃលទ្ធផលទាំងនោះ ដែលនឹងជួយអ្នកយ៉ាងខ្លាំងនៅពេលអ្នកប្រយុទ្ធជាមួយទិន្នន័យរបស់អ្នក។


### ការដោះស្រាយទិន្នន័យដែលខ្វះ

> **គោលបំណងរៀន:** នៅចុងផ្នែកនេះ អ្នកគួរតែដឹងពីរបៀបនិងពេលដែលត្រូវប្តូរឬដកតម្លៃ null ចេញពី DataFrames។

គំរូ Machine Learning មិនអាចដោះស្រាយទិន្នន័យដែលខ្វះដោយខ្លួនឯងបានទេ។ ដូច្នេះ មុនពេលផ្ញើទិន្នន័យចូលទៅក្នុងគំរូ យើងត្រូវតែដោះស្រាយតម្លៃខ្វះទាំងនេះ។

របៀបដោះស្រាយទិន្នន័យខ្វះមានការជួញដូរឆ្លាតវៃ មានអនុលោមអាស្រ័យលើវា អាចប៉ះពាល់ដល់វិភាគចុងក្រោយរបស់អ្នក និងលទ្ធផលក្នុងពិភពជាក់ស្តែង។

មានរបៀបសំខាន់ចម្បងពីរនៃការដោះស្រាយទិន្នន័យខ្វះ៖


1.   លុបជួរដែលមានតម្លៃខ្វះចេញ
2.   ប្តូរតម្លៃខ្វះដោយតម្លៃផ្សេងទៀត

យើងនឹងពិភាក្សាអំពីវិធីទាំងពីរនេះ និងអត្ថប្រយោជន៍និងគុណវិបត្តិរបស់ពួកវាជារយៈលម្អិត។


### ការលុបតម្លៃទទេ

បរិមាណទិន្នន័យដែលយើងផ្តល់ចូលទៅកាន់ម៉ូដែលរបស់យើងមានឥទ្ធិពលផ្ទាល់លើកម្រិតប្រសិទ្ធភាពរបស់វា។ ការលុបតម្លៃទទេមានន័យថាយើងកំពុងកាត់ចំនួនចំណុចទិន្នន័យ ហើយដូច្នេះក៏កាត់បន្ថយទំហំឯកសារទិន្នន័យផងដែរ។ ដូច្នេះ គឺមានការណែនាំឱ្យលុបជួរដេកដែលមានតម្លៃទទេនៅពេលដែលឯកសារទិន្នន័យមានទំហំធំ។

ករណីផ្សេងទៀតអាចជាជួរដេកឬជួរឈរមួយមានតម្លៃខ្វះខាតច្រើន។ បន្ទាប់មក វាអាចត្រូវបានលុបចោល ព្រោះវាមិននាំមកនូវតម្លៃច្រើនសម្រាប់វិភាគរបស់យើងទេ ដោយសារតែទិន្នន័យភាគច្រើនខ្វះសម្រាប់ជួរដេក/ជួរឈរនោះ។

លើសពីការកំណត់តម្លៃខ្វះ យ៉ាង pandas ផ្តល់នូវវិធីងាយស្រួលក្នុងការលុបតម្លៃទទេ ចេញពី `Series` និង `DataFrame`។ ដើម្បីឃើញវាក្នុងសកម្មភាព សូមត្រឡប់ទៅ `example3`។ មុខងារ `DataFrame.dropna()` ជួយក្នុងការលុបជួរដេកដែលមានតម្លៃទទេ។


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

សូមចំណាំថា នេះគួរត្រូវមើលទៅដូចជាផលចេញពី `example3[example3.notnull()]`។ ខុសគ្នាទៅនេះ គឺថា មិនបានបញ្ជាក់ត្រឹមតែតម្លៃដែលបានដាក់ម៉ាសទេ ប៉ុន្តែ `dropna` បានដកតម្លៃដែលខ្វះទាំងនោះចេញពី `Series` `example3`។

ដោយសារតែ DataFrames មានពីរ និងមិន មានទំហំវាត្រូវបានផ្តល់ជម្រើសច្រើនសម្រាប់ការដកចេញទិន្នន័យ។


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

,0,1,2
0,1.0,NaN,7
1,2.0,5.0,8
2,NaN,6.0,9


(តើអ្នកកត់សម្គាល់ទេថា pandas បានបម្លែងពីជួរឈរចំណួនពីរជា float ដើម្បីគាំទ្រ `NaN`?)

អ្នកមិនអាចលុបតម្លៃតែមួយចេញពី `DataFrame` ទេ ដូច្នេះអ្នកត្រូវតែលុបជួរដេកឬជួរឈរ ពេញលេញ។ អាស្រ័យតាមអ្វីដែលអ្នកកំពុងធ្វើ អ្នកអាចចង់ធ្វើមួយក្នុងចំណោមបាន ហើយដូច្នេះ pandas ផ្តល់ជម្រើសទាំងពីរឲ្យអ្នក។ ពីព្រោះនៅក្នុងវិទ្យាសាស្ត្រទិន្នន័យ ជួរឈរតំណាងឲ្យអថេរ ហើយជួរដេកតំណាងឲ្យការសង្កេត អ្នកតែងតែលុបជួរដេកនៃទិន្នន័យ; ការកំណត់លំនាំដើមសម្រាប់ `dropna()` គឺលុបជួរដេកទាំងអស់ដែលមានតម្លៃសូន្យណាមួយ៖


In [23]:
example4.dropna()

,0,1,2
1,2.0,5.0,8


ប្រសិនបើចាំបាច់ អ្នកអាចលុបតម្លៃ NA ពីជួរឈរ។ ប្រើ `axis=1` ដើម្បីធ្វើដូចនេះ៖


In [24]:
example4.dropna(axis='columns')

,2
0,7
1,8
2,9


សូមចំណាំថានេះអាចបណ្តាលឱ្យមានការបាត់បង់ទិន្នន័យច្រើនដែលអ្នកប្រហែលជាចង់រក្សាទុកពិសេសនៅក្នុងទិន្នន័យតូចៗ។ តើយ៉ាងដូចម្តេចបើអ្នកចង់បញ្ឈប់តែជួរដេកឬជួរឈរដែលមានតម្លៃទទេច្រើន ឬសូម្បីតែទាំងអស់គ្នាឲ្យបានបាត់បង់? អ្នកកំណត់ការដាក់តម្លៃនោះនៅក្នុង `dropna` ជាមួយប៉ារ៉ាម៉ែត្រ `how` និង `thresh`។

លំនាំដើមគឺ `how='any'` (បើអ្នកចង់ពិនិត្យផ្សេងទៀតដោយខ្លួនឯង ឬមើលថា តើវិធីសាស្ត្រផ្សេងទៀតមានអ្វីខ្លះ អ្នកអាចរត់ `example4.dropna?` ក្នុងកោដ្ឋិកូដមួយ)។ អ្នកអាចជ្រើសរើស `how='all'` ដើម្បីបញ្ឈប់តែជួរដេកឬជួរឈរដែលមានតម្លៃទទេទាំងអស់ប៉ុណ្ណោះ។ មកដោះស្រាយឧទាហរណ៍ `DataFrame` របស់យើងឲ្យធំជាងនេះដើម្បីឃើញការបញ្ជាក់វាកើតឡើងយ៉ាងដូចម្តេចនៅក្នុងលំហាត់បន្ទាប់។


In [25]:
example4[3] = np.nan
example4

,0,1,2,3
0,1.0,NaN,7,NaN
1,2.0,5.0,8,NaN
2,NaN,6.0,9,NaN


> ចំណុចសំខាន់ៗ៖  
1. ការបោះបង់តម្លៃ null គឺជាគំនិតល្អបើតែក្រុមទិន្នន័យធំប៉ុណ្ណោះ។  
2. ជួរឈរឬក៏កូឡុំទាំងមូលអាចត្រូវបានបោះបង់ ប្រសិនបើវាអត់មានទិន្នន័យភាគច្រើន។  
3. វិធីសាស្រ្ត `DataFrame.dropna(axis=)` ជួយក្នុងការកែចេញតម្លៃ null។ អេហ្ស៊ីង `axis` សម្គាល់ថាតើត្រូវបោះបង់ជួរឈរឬកូឡុំ។  
4. អេហ្ស៊ីង `how` ក៏អាចប្រើបានផងដែរ។ ដោយលំនាំដើម វាត្រូវបានកំណត់ជា `any`។ ដូច្នេះ វាបោះបង់តែនូវជួរឈរ/កូឡុំដែលមានតម្លៃ null យ៉ាងហោចណាស់មួយ។ វាអាចកំណត់ជា `all` ដើម្បីបញ្ជាក់ថា យើងនឹងបោះបង់តែនូវជួរឈរ/កូឡុំនោះទេដែលមានតម្លៃមួយចំនួនសុទ្ធ null។


### លំហាត់:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


ប៉ារ៉ាម៉ែត្រ `thresh` អនុញ្ញាតឱ្យអ្នកបញ្ជាការតាមលំដាប់លម្អិតជាងមុន៖ អ្នកកំណត់ចំនួនតម្លៃ *មិនមែនទទេ* ដែលជួរដេក ឬជួរឈរមួយត្រូវតែមានដើម្បីរក្សាទុក៖


In [27]:
example4.dropna(axis='rows', thresh=3)

,0,1,2,3
1,2.0,5.0,8,NaN


នៅទីនេះ ជួរដើមនិងចុងមួយត្រូវបានដាក់ទៅក្រោម ពីព្រោះវាមានតែពីរតម្លៃមិនទទេប៉ុណ្ណោះ។


### បំពេញតម្លៃ​ទទេ

ពេលខ្លះ វាអាចមានអត្ថប្រយោជន៍ក្នុងការបំពេញតម្លៃដែលខ្វះដោយតម្លៃដែលអាចមានភាពត្រឹមត្រូវ។ មានបច្ចេកទេសមួយចំនួនសម្រាប់បំពេញតម្លៃទទេ។ វិធីដំបូងគឺប្រើចំណេះដឹងអំពីដែនកំណត់ (ចំណេះដឹងអំពីប្រធានបទដែលទិន្នន័យភ្ជាប់) ដើម្បីប្រមាណតម្លៃដែលខ្វះ។

អ្នកអាចប្រើ `isnull` ដើម្បីធ្វើការបំពេញនេះនៅទីតាំងដើមបាន ប៉ុន្តែវាអាចពិបាក ជាពិសេសបើអ្នកមានតម្លៃច្រើនដែលត្រូវបំពេញ។ ដោយសារតែនេះគឺជាការងារពេញនិយមក្នុងវិទ្យាសាស្ត្រទិន្នន័យ pandas បានផ្តល់ជូន `fillna` ដែលត្រឡប់រូបមន្តមួយនៃ `Series` ឬ `DataFrame` ដែលមានតម្លៃខ្វះត្រូវបានប្តូរជាមួយតម្លៃដែលអ្នកជ្រើសរើស។ យើងមកបង្កើតឧទាហរណ៍ថ្មីមួយទៀតនៃ `Series` ដើម្បីមើលថាវាធ្វើការដូចម្តេចក្នុងការអនុវត្ត។


### ទិន្នន័យប្រភេទក្រុម(មិនមែនជាលេខ)
មុនគេ យើងមកពិចារណាអំពីទិន្នន័យមិនមែនជាលេខ។ ក្នុងសំណុំទិន្នន័យ យើងមានជួរឈរដែលមានទិន្នន័យប្រភេទក្រុម។ ឧទាហរណ៍ ភេទ, ត្រឹមត្រូវ ឬ មិនត្រឹមត្រូវ ល។ 

នៅក្នុងករណីភាគច្រើននេះ យើងជំនួសតម្លៃខ្វះដោយ `mode` របស់ជួរឈរ។ បើសិនជាយើងមានចំណុចទិន្នន័យ ១០០ និង ៩០ នាក់បាននិយាយថា ត្រឹមត្រូវ, ៨ នាក់បាននិយាយថា មិនត្រឹមត្រូវ ហើយ ២ នាក់មិនបានបំពេញ។ បន្ទាប់មក យើងអាចបំពេញ ២ នាក់ជាមួយតម្លៃ ត្រឹមត្រូវ ដោយគិតពីជួរឈរសរុប។

ម្ដងទៀត នៅទីនេះ យើងអាចប្រើចំណេះដឹងផ្នែកដែនដី។ យើងមកពិចារណាឧទាហរណ៍នៃការបំពេញដោយ mode។


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

,0,1,2
0,1,2,True
1,3,4,None
2,5,6,False
3,7,8,True
4,9,10,True


ឥឡូវនេះ សូមយើងស្វែងរកម៉ូដមុនពេលបំពេញតម្លៃ `None` ជាមួយម៉ូដ។


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

ដូច្នេះ យើងនឹងជំនួស None ជា True។


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

,0,1,2
0,1,2,True
1,3,4,True
2,5,6,False
3,7,8,True
4,9,10,True


ដូចដែលយើងអាចមើលឃើញ តម្លៃ null ត្រូវបានប្តូរជំនួស។ មិនចាំបាច់និយាយថា យើងអាចបានសរសេរអ្វីមួយណាមួយជំនួសឬ `'True'` ហើយវានឹងត្រូវបានប្តូរជំនួសផងដែរ។


### ទិន្នន័យគណនាផ្នែកចំនួន
ឥឡូវនេះ មកដល់ទិន្នន័យគណនាផ្នែកចំនួន។ នៅទីនេះ យើងមានមធ្យោបាយពីរដើម្បីជំនួសតម្លៃចាញ់ខ្វះ៖

1. ជំនួសជាមួយមេឌៀននៃជួរដេក
2. ជំនួសជាមួយមធ្យមនៃជួរដេក

យើងជំនួសជាមួយមេឌៀន ក្នុងករណីទិន្នន័យមានការបែកដេកជាមួយតម្លៃខុសៗគ្នា។ នេះសំរាប់មេឌៀនមានភាពរឹងមាំចំពោះតម្លៃខុសៗគ្នា។

ពេលដែលទិន្នន័យត្រូវបានធម្មតា (normalized) យើងអាចប្រើមធ្យមបាន ព្រោះនៅក្នុងករណីនោះ មធ្យមនិងមេឌៀននឹងមានតម្លៃជិតគ្នា។

ដំបូង យើងចាប់យកជួរឈរមួយដែលបែងចែកធម្មតា ហើយឲ្យយើងបំពេញតម្លៃខ្វះជាមួយមធ្យមនៃជួរឈរនោះ។


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

,0,1,2
0,-2.0,0,1
1,-1.0,2,3
2,NaN,4,5
3,1.0,6,7
4,2.0,8,9


មធ្យមភាគនៃជួរឈរនេះគឺ


In [33]:
np.mean(fill_with_mean[0])

0.0

បំពេញជាមួយค่า​មធ្យម


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

,0,1,2
0,-2.0,0,1
1,-1.0,2,3
2,0.0,4,5
3,1.0,6,7
4,2.0,8,9


ដូចដែលយើងអាចឃើញ តម្លៃដែលបាត់បង់បានត្រូវបានជំនួសដោយមធ្យមភាគរបស់វា។


ឥឡូវនេះ យើងចូរព្យាយាម dataframe មួយផ្សេងទៀត ហើយពេលនេះ យើងនឹងជំនួសតម្លៃ None ជាមួយមិត្ត​រដ្ឋកណ្តាលនៃកូឡុំ។


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

,0,1,2
0,-2,0.0,1
1,-1,2.0,3
2,0,NaN,5
3,1,6.0,7
4,2,8.0,9


មធ្យមភាគនៃជួរឈរទីពីរគឺ


In [36]:
fill_with_median[1].median()

4.0

បម្រះដោយម​ធ្យមភាគ


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

,0,1,2
0,-2,0.0,1
1,-1,2.0,3
2,0,4.0,5
3,1,6.0,7
4,2,8.0,9


ដូចដែលយើងឃើញ តម្លៃ NaN ត្រូវបានជំនួសដោយតម្លៃមធ្យមកណ្តាលនៃជួរឈរ។


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

អ្នកអាចបំពេញចន្លោះទទេទាំងអស់ដោយតម្លៃតែមួយបាន ដូចជារាប់ `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> ចំណុចសំខាន់:
1. ការបំពេញតម្លៃដែលខ្វះគួរត្រូវធ្វើពេលមានទិន្នន័យតិច ឬមានយុទ្ធសាស្រ្តក្នុងការបំពេញទិន្នន័យដែលខ្វះ។
2. ចំណេះដឹងដែនកំណាត់អាចត្រូវប្រើដើម្បីបំពេញតម្លៃដែលខ្វះដោយប្រហាក់ប្រហែលវា។
3. សម្រាប់ទិន្នន័យប្រភេទកាតេហ្គរីភិក, ភាគច្រើនតម្លៃដែលខ្វះត្រូវបានជំនួសជាមួយម៉ូដនៃជួរឈរ។
4. សម្រាប់ទិន្នន័យគណិតវិទ្យា, តម្លៃដែលខ្វះភាគច្រើនត្រូវបានបំពេញជាមួយមធ្យម (សម្រាប់ឃមnormalized datasets) ឬមធ្យមភាគពីរ (median) នៃជួរឈរ។


### អនុវត្តវិជ្ជា:


In [40]:
# What happens if you try to fill null values with a string, like ''?


អ្នកអាច **បំពេញតម្លៃ​ទំនេរ** ជាមុខមែន ដោយប្រើតម្លៃត្រឹមត្រូវចុងក្រោយដើម្បីបំពេញតម្លៃទំនេរ៖


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

អ្នកក៏អាច **បំពេញត្រឡប់** ដើម្បីផ្សព្វផ្សាយតម្លៃត្រឹមត្រូវបន្ទាប់ទៅក្រោយត្រឡប់មកដើម្បីបំពេញតម្លៃទទេបានផងដែរ៖


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

ដូចដែលអ្នកអាចរើសគំនិតបាន វានៅតែមើលទៅដូចគ្នាជាមួយ DataFrames ប៉ុន្តែអ្នកក៏អាចបញ្ជាក់ `axis` មួយដែលត្រូវបំពេញតម្លៃទទេ។


In [43]:
example4

,0,1,2,3
0,1.0,NaN,7,NaN
1,2.0,5.0,8,NaN
2,NaN,6.0,9,NaN


In [44]:
example4.fillna(method='ffill', axis=1)

,0,1,2,3
0,1.0,1.0,7.0,7.0
1,2.0,5.0,8.0,8.0
2,NaN,6.0,9.0,9.0


សូមចំណាំថា ពេលដែលមូល្យតម្លៃមុនមិនមានសម្រាប់បំពេញទៅមុខ តម្លៃទទេនោះនឹងនៅតែមិនប្រែប្រួល។


### ព្យាយាម:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


អ្នកអាចមានភាពច្នៃប្រឌិតអំពីវិធីដែលអ្នកប្រើ `fillna` បាន។ ឧទាហរណ៍ មកមើល `example4` ម្តងទៀត ប៉ុន្តេលើកនេះ យើង filling តម្លៃខ្វះជាមួយមធ្យមផ្សំទាំងអស់ក្នុង `DataFrame`:


In [46]:
example4.fillna(example4.mean())

,0,1,2,3
0,1.0,5.5,7,NaN
1,2.0,5.0,8,NaN
2,1.5,6.0,9,NaN


សូមស្ដាប់ថា កាលុំនី 3 នៅតែមិនមានតម្លៃ៖ ទិសដៅលំនាំដើមគឺបំពេញតម្លៃតាមជួរដេក។

> **ចំណាប់អារម្មណ៍៖** មានវិធីជាច្រើនក្នុងការដោះស្រាយតម្លៃខ្វះនៅក្នុងសំណុំទិន្នន័យរបស់អ្នក។ គោលការណ៍ជាក់លាក់ដែលអ្នកប្រើ (បែកចេញ ព្យួរប្រែប្លែង ឬហើយថា វិធីដែលអ្នកប្តូរពួកវា) គួរតែត្រូវបានកំណត់ដោយលក្ខណៈពិសេសនៃទិន្នន័យនោះ។ អ្នកនឹងអភិវឌ្ឍការយល់ដឹងល្អប្រសើរច្រើនពីរបៀបដោះស្រាយតម្លៃខ្វះនៅពេលដែលអ្នកដោះស្រាយ និងទាក់ទងជាមួយសំណុំទិន្នន័យ។


### ការអ៊ិនកូដទិន្នន័យប្រភេទផ្នែក

ម៉ូដែលរៀនម៉ាស៊ីនគ្រាន់តែដំណើរការលេខនិងទិន្នន័យគណនាត្រឹមតែ។ វា​នឹងមិន​អាច​បំបែក​ភាពខុសគ្នារវាង បាទ និង ទេ ទេ ប៉ុន្តែវានឹងអាចបំបែកចន្លោះ 0 និង 1 តែប៉ុណ្ណោះ។ ដូច្នេះ បន្ទាប់ពីបំពេញតម្លៃខ្វះ យើងត្រូវតែអ៊ិនកូដទិន្នន័យប្រភេទផ្នែកទៅទ្រង់ទ្រង់លេខសម្រាប់ម៉ូដែលយល់។

ការអ៊ិនកូដអាចធ្វើបានពីរបៀប។ យើងនឹងពិភាក្សាពួកវាតទៅមុខ។


**ការអ៊ិនកូដស្លាក**


ការអ៊ិនកូដស្លាកគឺជាការបំលែងប្រភេទគ្រប់មួយទៅជាលេខ។ ឧទាហរណ៍ មេីលយើងមានទិន្នន័យអ្នកធ្វើដំណើរការយន្តហោះ ហើយមានជួរឈរមួយដែលមានថ្នាក់របស់ពួកគេក្នុងចំណោម ['business class', 'economy class','first class']។ ប្រសិនបើអ៊ិនកូដស្លាកត្រូវបានអនុវត្តលើនេះ វានឹងប្តូរទៅជា [0,1,2]។ មកមេីលឧទាហរណ៍តាមកូដ។ ពីព្រោះយើងនឹងរៀន `scikit-learn` ក្នុងកំណត់ត្រាចេញក្រោយ យើងមិនប្រើវាដោយផ្ទាល់នៅទីនេះទេ។


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

,ID,class
0,10,business class
1,20,first class
2,30,economy class
3,40,economy class
4,50,economy class
5,60,business class


ដើម្បីអនុវត្តការវិភាគស្លាកនៅជួរឈរទីមួយ យើងត្រូវតែពណ៌នាការបម្លែងពីថ្នាក់នីមួយៗទៅជាចំនួនមួយជាមុនសិន មុនពេលប្ដូរឡើងវិញ។


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

,ID,class
0,10,0
1,20,2
2,30,1
3,40,1
4,50,1
5,60,0


ដូចដែលយើងអាចមើលឃើញ លទ្ធផលផ្ទ keluម្រប់គ្នានឹងអ្វីដែលយើងគិតថានឹងកើតឡើង។ ដូច្នេះ ពេលណាមួយយើងប្រើការបំលែងស្លាក? ការបំលែងស្លាកត្រូវបានប្រើនៅក្នុងករណីណាមួយ ឬទាំងពីរនៃករណីខាងក្រោម៖
1. នៅពេលដែលចំនួនប្រភេទធំ
2. នៅពេលដែលប្រភេទមានលំដាប់។


**ONE HOT ENCODING**

ប្រភេទកូដមួយទៀតគឺ One Hot Encoding។ ក្នុងប្រភេទកូដនេះ ប្រភេទនីមួយៗនៃជួរឈរត្រូវបានបន្ថែមជាជួរឈរផ្សេងបង្កើតឡើង ហើយចំណុចទិន្នន័យនិមួយៗនឹងទទួលបានលេខ 0 ឬ 1 ដោយផ្អែកលើថាតើវាមានប្រភេទនោះ។ ដូច្នេះ ប្រសិនបើមានប្រភេទខុសៗគ្នា n ប្រភេទ នោះជួរឈរ n នឹងត្រូវបានបូកបន្ថែមចូលទៅក្នុងតារាងទិន្នន័យ។

ឧទាហរណ៍ យើងសន្និដ្ឋានថាតែមួយថ្នាក់យន្តហោះដូចគ្នា។ ប្រភេទគឺជា៖ ['business class', 'economy class','first class']។ ដូច្នេះ ប្រសិនបើយើងអនុវត្ត One Hot Encoding ជួរឈរទាំងបីខាងក្រោមនឹងត្រូវបានបន្ថែមទៅក្នុងសំណុំទិន្នន័យ៖ ['class_business class','class_economy class','class_first class']។


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

,ID,class
0,10,business class
1,20,first class
2,30,economy class
3,40,economy class
4,50,economy class
5,60,business class


យើងអនុវត្តការអ៊ិនកូដឯកតាមួយលើជួរឈរទី ១


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

,ID,class_business class,class_economy class,class_first class
0,10,1,0,0
1,20,0,0,1
2,30,0,1,0
3,40,0,1,0
4,50,0,1,0
5,60,1,0,0


ជួរឈរ​កូដ​ជា​មួយ​តែមួយ​មួយគឺ​មាន​តម្លៃ 0 ឬ 1 ដែល​បញ្ជាក់​ថា​ចំណាត់ថ្នាក់​នោះ​មាន​សម្រាប់​ចំណុច​ទិន្នន័យ​នោះ​ឬ​អត់។


តើយើងប្រើការវាយលេខ one hot encoding ពេលណា? ការវាយលេខ one hot encoding ត្រូវបានប្រើនៅពេលណាមួយ ឬទាំងពីររឿងខាងក្រោមនេះ៖

1. ពេលដែលចំនួនប្រភេទនិងទំហំឃ្លាំងទិន្នន័យមានតូច។
2. ពេលដែលប្រភេទទាំងអស់មិនមានលំដាប់ពិសេសណាមួយ។


> ពិន្ទុចម្បងដែលគួរចងចាំ៖
1. ការវាយកូដត្រូវបានធ្វើឡើងដើម្បីបំលែងទិន្នន័យដែលមិនមែនជាចំនួនទៅជាទិន្នន័យជាចំនួន។
2. មានប្រភេទការវាយកូដពីរប្រើ៖ ការវាយកូដស្លាក និងការវាយកូដ One Hot ដែលទាំងពីរអាចអនុវត្តបានដោយផ្អែកលើតម្រូវការរបស់សំណុំទិន្នន័យ។


## លុបយកទិន្នន័យមេរៀនមកវិញ

> **គោលបំណងការសិក្សា៖** នៅចុងបញ្ចប់នៃផ្នែកបន្ទាប់នេះ អ្នកគួរតែមានភាពស៑កស្អិតក្នុងការបញ្ជាក់ និងលុបចោលតម្លៃមេរៀនពី DataFrames។

ក្រៅពីទិន្នន័យដែលខ្វះ អ្នកនឹងជួបប្រទៈនឹងទិន្នន័យមេរៀនជាញឹកញាប់ក្នុងឯកសារទិន្នន័យពិតប្រាកដ។ ជាថាស័ក្តិ pandas ផ្តល់នូវវិធីងាយស្រួលក្នុងការស្វែងរក និងលុបចោលបញ្ចូលព័ត៌មានមេរៀន។


### ការកំណត់អត្តសញ្ញាណចម្លង: `duplicated`

អ្នកអាចរកឃើញតម្លៃចម្លងបានយ៉ាងងាយស្រួលដោយប្រើវិធីសាស្ត្រ `duplicated` ក្នុង pandas ដែលផ្តល់ជាមាសប៊៊ូលបង្ហាញថាតើចំណងជើងមួយនៅក្នុង `DataFrame` ជាចម្លងនៃចំណងជើងមួយដែលមានមុនធ្លាប់ឬអត់។ តោះបង្កើត `DataFrame` មួយផ្សេងទៀតដើម្បីមើលវានៅក្នុងសកម្មភាព។


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

,letters,numbers
0,A,1
1,B,2
2,A,1
3,B,3
4,B,3


In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### ការលុបចោលចម្លងទិន្នន័យ៖ `drop_duplicates`
`drop_duplicates` គឺត្រឹមតែបម្លងទិន្នន័យមួយច្បាស់សម្រាប់ដែលមានតម្លៃ `duplicated` ទាំងអស់ជា `False` ៖


In [54]:
example6.drop_duplicates()

,letters,numbers
0,A,1
1,B,2
3,B,3


ទាំង `duplicated` និង `drop_duplicates` តែងតែគិតគូសបញ្ជាក់ទៅលើព្រមគ្រប់ជួរឈរ ប៉ុន្តែអ្នកអាចបញ្ជាក់ថាពួកវាគ្រោងតែពិនិត្យតែផ្នែកខ្លះនៃជួរឈរនៅក្នុង `DataFrame` របស់អ្នកបាន៖


In [55]:
example6.drop_duplicates(['letters'])

,letters,numbers
0,A,1
1,B,2


> **សេចក្តីសង្ខេប:** ការដកចេញទិន្នន័យមួយរំពាស់គឺជាផ្នែកសំខាន់របស់គម្រោងវិទ្យាសាស្ត្រទិន្នន័យភាគច្រើន។ ទិន្នន័យមួយរំពាស់អាចផ្លាស់ប្តូរប្រាក់លទ្ធផលនៃការវិភាគរបស់អ្នក និងផ្តល់លទ្ធផលមិនត្រឹមត្រូវជូនអ្នក!


## ការត្រួតពិនិត្យគុណភាពទិន្នន័យនៅក្នុងពិភពជាក់ស្តែង

> **គោលដៅការសិក្សា:** នៅចុងផ្នែកនេះ អ្នកគួរតែមានភាពស្រួលក្នុងការសម្រេចឲ្យបានថា ពេលឃើញបញ្ហាគុណភាពទិន្នន័យជាក់ស្តែងដូចជា តម្លៃប្រភេទមានការមិនសំរបសំរួល តម្លៃលេខវិភាគអាសកម្ម (outliers) និងអ្វីដែលធ្វើរចនាសម្ព័ន្ធទិន្នន័យត្រូវបញ្ចេញតាមបែបផែននៃការភច្ចរម្រៃ។

ទោះបីជាតម្លៃខ្វះនិងកំណត់ត្រាការចម្លងជាក់លាក់គឺជាបញ្ហាទូទៅក៏ដោយ មូលដ្ឋានទិន្នន័យនៅក្នុងពិភពជាក់ស្តែងភាគច្រើនផ្ទុកបញ្ហាលំបាកជាងនេះ៖

1. **តម្លៃប្រភេទដែលមិនសំរបសំរួល**៖ ប្រភេទដដែលដែលមានការវាយអក្សរផ្សេងៗគ្នា (ដូចជា "USA", "U.S.A", "United States")
2. **តម្លៃលេខវិភាគអសកម្ម**៖ គន្លងកំណត់ខ្ពស់ឬទាបខ្លាំងដែលបង្ហាញកំហុសបញ្ចូលទិន្នន័យ (ដូចជា អាយុ = 999)
3. **ជួរដូចគ្នា قريب**៖ កំណត់ត្រាដែលតំណាងឲ្យអEntityដូចគ្នា តែមានការប្រែប្រួលតិចតួច

ចូរយើងស្វែងយល់ពីវិធីសាស្រ្តក្នុងការរកឃើញ និងដោះស្រាយបញ្ហាទាំងនេះ។


### បង្កើតឯកសារទិន្នន័យ "ស្អំណោច" ឧទាហរណ៍

ជំហ៊តដំបូង មកបង្កើតឯកសារទិន្នន័យឧទាហរណ៍មួយដែលមានប្រភេទបញ្ហាដែលយើងជួបប្រទៈជាញឹកញាប់ក្នុងទិន្នន័យពិត៖


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. ការសម្គាល់តម្លៃចំណាត់ថ្នាក់មិនត្រឹមត្រូវ

សូមដាក់ចំណាំថា​កូឡុំ `country` មានការបង្ហាញច្រើនប្រភេទសម្រាប់ប្រទេសដដែល។ តោះយើងសម្គាល់ពីភាពមិនសមរម្យទាំងនេះ៖


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### ការធ្វើឱ្យតម្លៃប្រភេទសរុបស្មើគ្នា

យើងអាចបង្កើតការចងក្រងដើម្បីធ្វើឱ្យតម្លៃទាំងនេះស្មើគ្នា។ វិធីសាមញ្ញមួយ​គឺបំលែងទៅជាតួអក្សរតូច ហើយបង្កើតវចនានុក្រមចងក្រង:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**ជម្រើសមួយផ្សេងទៀត៖ ការប្រើប្រាស់ការចំរូងស្រដៀងគ្នា**

សម្រាប់ករណីស្មុគស្មាញជាងនេះ យើងអាចប្រើការចំរូងខ្សែអក្សរស្រដៀងគ្នាជាមួយបណ្ណាល័យ `rapidfuzz` ដើម្បីរកឃើញខ្សែអក្សរស្រដៀងគ្នា secara otomatis:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. ការរកឃើញតម្លៃចំនួនអសកម្ម (Outliers)

មើលទៅកូឡাম `age` យើងមានតម្លៃខុសប្រក្រតីមួយចំនួនដូចជា 199 និង -5។ អ្នកមកប្រើវិធីសាស្រ្តស្ថិតិដើម្បីរកឃើញអត្រាចេញនេះ។


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### ការប្រើវិធីសាស្រ្ត IQR (ចន្លោះចំណែកបួន) 

វិធីសាស្រ្ត IQR គឺជាយុទ្ធសាស្រ្តស្ថិតិរឹងមាំសម្រាប់ការរកចេញចំណុចបម្រែង ដែលមានភាពមិនទំនាក់ទំនងច្រើនទៅតម្លៃខ្ពស់ខ្នាញ់:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### ការប្រើវិធាន Z-Score

វិធីសាស្រ្ត Z-score កំណត់អ្នកចេញពីដែនលើមូលដ្ឋានលើវិធីសាស្រ្តធម្មតាម៉ាប់ពីម៉ូតុងមធ្យម:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### ការដោះស្រាយចំណុចចម្លែក

បន្ទាប់ពីរកឃើញ រួចហើយ ចំណុចចម្លែកអាចត្រូវបានដោះស្រាយដោយវិធីខ្លះៗដូចខាងក្រោម៖
1. **លុបចោល**៖ លុបជួរដេកមានចំណុចចម្លែក (បើវាជាកំហុស)
2. **កំណត់កម្រិត**៖ ផ្លាស់ទីជាមួយតម្លៃកំណត់ស្វែង
3. **ជំនួសជាមួយ NaN**៖ គិតថាជាទិន្នន័យខ្វះ ហើយប្រើបច្ចេកទេសបញ្ចូលតម្លៃបាត់បង់
4. **រក្សាទុក**៖ ប្រសិនបើវាជាតម្លៃអញ្ញើញស្រឡាយត្រឹមត្រូវ


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. ការរកឃើញជួរដូចគ្នា​ជិតៗ

សូមកត់សម្គាល់ថាឯកសារទិន្នន័យរបស់យើងមានការចុះបញ្ជីច្រើនសម្រាប់ "John Smith" ដែលមានតម្លៃខុសគ្នារបៀបស្រាល។ យើងនឹងស្វែងរកចម្លើយដែលអាចជាអត្រាចម្លងដោយផ្អែកលើភាពស្រដៀងនៃឈ្មោះ។


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### ការស្វែងរកចម្លើយជិតដូចគ្នាដោយប្រើការភាពរសជា

សម្រាប់ការស្វែងរកចម្លើយតែបន្តិចទៅមុខជាងនេះ យើងអាចប្រើការភាពរសជាដើម្បីស្វែងរកឈ្មោះដែលស្រដៀងគ្នា:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### ការដោះស្រាយចម្លងមួយៗ

ពេលដែលបានសម្គាល់ហើយ អ្នកត្រូវសម្រេចចិត្តពីវិធីដែលត្រូវដោះស្រាយចម្លងមួយៗៈ
1. **រក្សាទុកករណីជាលំដាប់ដំបូង**: ប្រើ `drop_duplicates(keep='first')`
2. **រក្សាទុកករណីជាលំដាប់ចុងក្រោយ**: ប្រើ `drop_duplicates(keep='last')`
3. **សរុបព័ត៌មាន**: បញ្ចូលព័ត៌មានពីជួរដេកចម្លង
4. **ពិនិត្យដោយដៃ**: សញ្ញាសម្រាប់ការពិនិត្យដោយមនុស្ស


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### សេចក្ដីសង្ខេប៖ ស៊ុមដំណើរការសម្អាតទិន្នន័យពេញលេញ

មកធ្វើវាជាស៊ុមដំណើរការសម្អាតទិន្នន័យទាំងមូលគ្នា៖


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 លំហាត់ប្រកួត

ឥឡូវនេះជំពាក់អ្នក! ខាងក្រោមនេះជាឈុតទិន្នន័យថ្មីមួយដែលមានបញ្ហាគុណភាពច្រើន។ តើអ្នកអាច៖

1. កំណត់បញ្ហាទាំងអស់នៅក្នុងជួរនេះ
2. សរសេរកូដដើម្បីសម្អាតបញ្ហា​នីមួយៗ
3. បន្ថែមជួរដែលបានសម្អាតទៅកាន់សំណុំទិន្នន័យ

នេះជាទិន្នន័យដែលមានបញ្ហា៖


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### ចំណុចសំខាន់ៗ

1. **ប្រភេទដែលមិនស្របគ្នា** ជារឿងធម្មតានៅក្នុងទិន្នន័យពិភពលោកពិត។ តែងតែពិនិត្យតម្លៃពិសេសៗ ហើយស្តង់ដាឌ័រអ្នកប្រើប្រាស់ដោយប្រើផែនទីឬការប្រៀបធៀបដូចជា fuzzy matching។

2. **ចំណុចលើស** អាចមានឥទ្ធិពលយ៉ាងខ្លាំងលើការវិភាគរបស់អ្នក។ ប្រើចំណេះដឹងស្រុកជាមួយវិធីសាស្រ្តស្ថិតិបទ (IQR, Z-score) ដើម្បីរកតួអង្គទាំងនេះ។

3. **ចម្លងប្រហែល** គឺកាន់តែពិបាករកឃើញជាងចម្លងត្រឹមត្រូវ។ គិតពីការប្រើ fuzzy matching និងធម្មតាតម្លៃទិន្នន័យ (ធ្វើអក្សរតូច, ដកចេញដ៏ចាល់លើកន្លែងទំហឹង) ដើម្បីរកឃើញវា។

4. **ការសម្អាតទិន្នន័យគឺជាការធ្វើជាការបន្តបន្ទាប់**។ អ្នកប្រហែលត្រូវការប្រើបច្ចេកទេសជាច្រើន ហើយពិនិត្យលទ្ធផលមុនពេលបញ្ចប់សំណុំទិន្នន័យសម្អាតរបស់អ្នក។

5. **ឯកសារសេចក្តីសម្រេចរបស់អ្នក**។ រក្សាតាមដាននូវជំហានសម្អាតដែលអ្នកបានអនុវត្ត និងហេតុផល, ព្រោះវាសំខាន់សម្រាប់ការបង្កើតឡើងវិញ និងភាពច្បាស់លាស់។

> **ការអនុវត្តដ៏ល្អបំផុត:** តែងតែរក្សាចម្លងនៃទិន្នន័យ "ស្អាត" ដើមរបស់អ្នក។ មិនត្រូវបិទលើឯកសារទិន្នន័យប្រតិបត្តិរបស់អ្នកទេ - បង្កើតកំណែស្អាតដែលមានការកំណត់ឈ្មោះច្បាស់លាស់ដូចជា `data_cleaned.csv` ។


---

<!-- CO-OP TRANSLATOR DISCLAIMER START -->
**ការបដិសេធ**៖  
ឯកសារនេះត្រូវបានបម្លែងភាសាដោយប្រើសេវាកម្មបកប្រែ AI [Co-op Translator](https://github.com/Azure/co-op-translator)។ ខណៈពេលដែលយើងខំប្រឹងប្រែងឲ្យបានត្រឹមត្រូវ សូមយល់ថាការបកប្រែដោយស្វ័យប្រវត្តិអាចមានកំហុសឬភាពមិនត្រឹមត្រូវ។ ឯកសារដើមជាភាសាមាតុភូមិគួរត្រូវបានគេពិចារណាថាជា ប្រភពផ្លូវការជាផ្លូវការ។ សម្រាប់ព័ត៌មានដ៏សំខាន់ សូមផ្តល់លទ្ធភាពការបកប្រែដោយមនុស្សជំនាញ។ យើងមិនទទួលខុសត្រូវចំពោះការយល់ច្រឡំ ឬការបកស្រាយខុសប្រកបដោយការប្រើប្រាស់ការបកប្រែនេះទេ។
<!-- CO-OP TRANSLATOR DISCLAIMER END -->
